lib/src/geometry/collision_detection.dart

```diff
@@ -0,0 +1,78 @@
+import 'dart:collection';
+
+import '../components/mixins/collidable.dart';
+import 'line_segment.dart';
+import 'shape.dart';
+import '../../extensions.dart';
+
+/// Check whether any [Collidable] in [collidables] collide with each other
+/// or [screenSize] (if defined), and call callbacks accordingly
+void collisionDetection(List<Collidable> collidables) {
+  for (int x = 0; x < collidables.length; x++) {
+    final collidableX = collidables[x];
+    for (int y = x + 1; y < collidables.length; y++) {
```

Comment:
both fors here can be for in

Reply:
They can't, since the second one is dependent on the current index of the first, so that we don't do checks between components that have already been checked against each other.

Reply:
ah didn't noticed that!